apiserver/facades/agent/secretsmanager/access.go

```diff
@@ -26,36 +29,71 @@ func (s *SecretsManagerAPI) hasRole(uri *coresecrets.URI, entity names.Tag, role
 	return err == nil && hasRole.Allowed(role)
 }
 
-func (s *SecretsManagerAPI) canManage(uri *coresecrets.URI, entity names.Tag) bool {
-	// Manage access is granted on a per app basis;
-	// Any unit from the allowed app can manage.
-	authAppName := authTagApp(entity)
-	if authAppName == "" {
-		return false
+// canManage checks that the authenticated caller can manage the secret, and returns a
+// token to ensure leadership if that is required; ie if the request is for a secret
+// owned by an application, the entity must be the unit leader.
+func (s *SecretsManagerAPI) canManage(uri *coresecrets.URI) (leadership.Token, error) {
+	// TODO(wallyworld) - remove auth tag kind check when podspec charms are gone.
+	if s.authTag.Kind() == names.UnitTagKind && s.hasRole(uri, s.authTag, coresecrets.RoleManage) {
+		return successfulToken{}, nil
 	}
-	app := names.NewApplicationTag(authAppName)
-	return s.hasRole(uri, app, coresecrets.RoleManage)
+	if s.authTag.Kind() != names.ApplicationTagKind {
+		return s.leadershipToken()
+	}
+	appName := authTagApp(s.authTag)
+	appTag := names.NewApplicationTag(appName)
+	if s.hasRole(uri, appTag, coresecrets.RoleManage) {
+		return successfulToken{}, nil
+	}
+	return nil, apiservererrors.ErrPerm
 }
 
+// canRead returns true if the specified entity can read the secret.
 func (s *SecretsManagerAPI) canRead(uri *coresecrets.URI, entity names.Tag) bool {
-	if s.canManage(uri, entity) {
-		return true
-	}
 	// First try looking up unit access.
 	hasRole, _ := s.secretsConsumer.SecretAccess(uri, entity)
 	if hasRole.Allowed(coresecrets.RoleView) {
 		return true
 	}
-	// If no unit level access granted, try for the app.
-	authAppName := authTagApp(entity)
-	if authAppName == "" {
-		return false
-	}
-	app := names.NewApplicationTag(authAppName)
-	hasRole, _ = s.secretsConsumer.SecretAccess(uri, app)
+	// TODO(wallyworld) - remove when podspec charms are gone.
+	appName := authTagApp(s.authTag)
+	hasRole, _ = s.secretsConsumer.SecretAccess(uri, names.NewApplicationTag(appName))
 	return hasRole.Allowed(coresecrets.RoleView)
 }
 
 func (s *SecretsManagerAPI) isSameApplication(tag names.Tag) bool {
 	return authTagApp(s.authTag) == authTagApp(tag)
 }
+
+// ownerToken returns a token used to determine if the specified entity
+// is owned by the authenticated caller.
+func (s *SecretsManagerAPI) ownerToken(ownerTag names.Tag) (leadership.Token, error) {
+	// A unit can create a secret so long as the
+	// secret owner is that unit's app.
+	// TODO(wallyworld) - remove auth tag kind check when podspec charms are gone.
+	if s.authTag.Kind() == names.ApplicationTagKind || s.authTag.Kind() == ownerTag.Kind() {
```

Comment:
`s.authTag.Id() == ownerTag.Id()`?

Reply:
yeah. also needed to move the app name check up to the top